pennylane/gradients/finite_difference.py

```diff
@@ -0,0 +1,258 @@
+
+    Returns:
+        array[float]: a ``(2, N)`` array. The first row corresponds to the
+        coefficients, and the second row corresponds to the shifts.
+
+    **Example**
+
+    >>> finite_diff_stencil(n=1, order=1, form="forward")
+    array([[-1.,  1.],
+           [ 0.,  1.]])
+
+    The first row corresponds to the coefficients, and the second corresponds
+    to the shifts. For example, this results in the linear combination:
+
+    .. math:: \frac{-y(x_0) + y(x_0 + h)}{h}
+
+    where :math:`h` is the finite-difference step-size.
+
+    More examples of finite-difference stencils:
+
+    >>> finite_diff_stencil(n=1, order=2, form="center")
+    array([[-0.5,  0.5],
+           [-1. ,  1. ]])
+    >>> finite_diff_stencil(n=2, order=2, form="center")
+    array([[-2.,  1.,  1.],
+           [ 0., -1.,  1.]])
+    """
+    if n < 1 or not isinstance(n, int):
+        raise ValueError("Derivative order n must be a positive integer.")
+
+    if order < 1 or not isinstance(order, int):
+        raise ValueError("Approximation order must be a positive integer.")
+
+    num_points = order + 2 * np.floor((n + 1) / 2) - 1
+    N = num_points + 1 if n % 2 == 0 else num_points
+
+    if form == "forward":
+        shifts = np.arange(N, dtype=np.float64)
+
+    elif form == "backward":
+        shifts = np.arange(-N + 1, 1, dtype=np.float64)
+
+    elif form == "center":
+        if order % 2 != 0:
+            raise ValueError("Centered finite-difference requires an even order.")
+
+        N = num_points // 2
+        shifts = np.arange(-N, N + 1, dtype=np.float64)
+
+    else:
+        raise ValueError(f"Unknown form {form}. Must be one of 'forward', 'backward', 'center'.")
+
+    A = shifts ** np.arange(len(shifts)).reshape(-1, 1)
+    b = np.zeros_like(shifts)
+    b[n] = factorial(n)
+    coeffs = np.linalg.solve(A, b)
+
+    stencil = np.stack([coeffs, shifts])
+    stencil[0, np.abs(stencil[0, :]) < 1e-10] = 0
+    stencil = stencil[:, ~np.all(stencil == 0, axis=0)]
+    stencil = stencil[:, np.argsort(np.abs(stencil)[1])]
+    return stencil
+
+
+def generate_shifted_tapes(tape, idx, shifts, multipliers=None):
+    r"""Generate a list of tapes where the corresponding trainable parameter
+    index has been shifted by the values given.
+
+    Args:
+        tape (.QuantumTape): input quantum tape
+        idx (int): trainable parameter index to shift the parameter of
+        shifts (Sequence[float or int]): sequence of shift values
+        multipliers (Sequence[float or int]): Sequence of multiplier values to
+            scale the paraameter by. If not provided, the parameter will
+            not be scaled.
+
+    Returns:
+        list[QuantumTape]: List of quantum tapes. Each tape has parameter
+        ``idx`` shifted by consecutive values of ``shift``. The length
+        of the returned list of tapes will match the length of ``shifts``.
+    """
+    params = qml.math.stack(tape.get_parameters())

```

Comment:
Just curious: this is always possible, because tape parameters are stored in a list of scalar tensors, right?

This may be something we have to reassess in the future if we want to use speedups in simulations stemming from the fact that we apply operations to full tensors ("tensor batching")...

Reply:
> Just curious: this is always possible, because tape parameters are stored in a list of scalar tensors, right?

Yes, it is a simplification I made last year which led to a lot cleaner code, but makes me worried for future compatibility...